packages/puppeteer-core/src/cdp/CdpSession.ts

```diff
@@ -140,22 +144,23 @@ export class CdpCDPSession extends CDPSession {
    * won't emit any events and can't be used to send messages.
    */
   override async detach(): Promise<void> {
-    if (!this.#connection) {
+    if (this.#closed) {
       throw new Error(
         `Session already detached. Most likely the ${this.#targetType} has been closed.`,
       );
     }
     await this.#connection.send('Target.detachFromTarget', {
       sessionId: this.#sessionId,
     });
+    this.#detached = true;
   }
 
   /**
    * @internal
    */
-  _onClosed(): void {
+  onClosed(): void {
     this.#callbacks.clear();
-    this.#connection = undefined;
```

Comment:
What was the rationale to introduce behavioral breaking change of `connection()` not returning `undefined` anymore after `_onClosed()` and at the same time make `get #closed()` private so that the state can not be accessed from the outside?

Reply:
I firmly believe this PR should've included also change alike https://github.com/puppeteer/puppeteer/pull/13613

Reply:
It was not intentional, but the there is no contract that enforces this behavior. 
If you were using this this was like using any other heuristic. 
As for why the close getter is private this is a private class so even if was not it would not be part of the public API.

Reply:
There's indeed no contract that enforces this behavior, however the `CDPSession` that `CdpCDPSession` implements is public, and `connection()` is not marked as internal (well, no explicitly public either).